client/src/GameState.ts

```diff
@@ -64,31 +70,131 @@ export class GameState {
         });
     }
+
+    public moveIfCan(
+        movement: (tetro: Tetromino) => TetrominoLookahead
+    ): boolean {
+        // look-ahead for the next tetromino state after movement
+        const lookahead = movement(this.currentTetromino);
+
+        if (
+            this.overlapWithBoard(lookahead) ||
+            this.overlapWithPlayers(lookahead, this.otherTetrominoes)
+        ) {
+            return false;
+        }
+
+        this.currentTetromino.updateFromLookahead(lookahead);
+        return true;
+    }
+
+    /**
+     * check against static board, see if newTetro is overlapping with static monominoes placed on board
+     * @returns boolean - if `newTetro` overlaps with any blocks on the board
+     */
+    private overlapWithBoard(lookahead: TetrominoLookahead): boolean {
+        return lookahead.tiles.some(([row, col]) => {
+            return this.board[row][col] != null;
+        });
+    }
+
+    /**
+     * check against other players' tetrominoes: see if newTetro is overlapping with other existing players
+     * @returns boolean - if `newTetro` overlaps with any of the `tetrominoes`
+     */
+    private overlapWithPlayers(
+        lookahead: TetrominoLookahead,
+        tetrominoes: Array<Tetromino>
+    ): boolean {
+        return tetrominoes.some((tetro) => {
+            // if the bounding boxes are more than the-maximum-"manhattan distance" away
+            const isFarAway =
+                Math.abs(tetro.position[0] - lookahead.position[0]) +
+                    Math.abs(tetro.position[1] - lookahead.position[1]) >
+                8; // 8 is twice the max width of a tetromino bounding box
+            if (isFarAway) {
+                return false;
+            }
+
+            // if newTetro has overlapping monominoes with those tetro
+            return tetro.tiles.some(([row, col]) => {
+                return lookahead.tiles.some(
+                    ([newRow, newCol]) => newRow === row && newCol === col
+                );
+            });
+        });
+    }
+
+    /**
+     * Returns if `tetro` is adjacent to any of the tetrominoes in the `tetrominoes` list
+     * @param tetro - The tetromino that is supposedly "currentTetromino", but can be any tetromino to be tested
+     * @param tetrominoes - The array of tetrominoes to be scanned through and looked for adjacentness
+     */
+    private adjacentWithPlayers(
+        tetro: Tetromino,
+        tetrominoes: Array<Tetromino>
+    ): boolean {
+        // No out of bounds check because overlapWithPlayers doesn't index board
+        const expandedSet = new Set(tetro.tiles);
+        tetro.tiles.forEach(([row, col]) => {
+            expandedSet.add([row + 1, col]);
+            expandedSet.add([row - 1, col]);
+            expandedSet.add([row, col + 1]);
+            expandedSet.add([row, col - 1]);
+        });
+
+        const lookahead = tetro.toTetrominoLookahead();
+        lookahead.tiles = Array.from(expandedSet);
```

Comment:
Just testing this in my browser console now, I don't think set actually removes duplicates here unfortunately.

Reply:
Nice catch... I feel the performance improvement is little, but I implemented a dedup version anyways.